src/typechecker.erl

```diff
@@ -1689,33 +1689,48 @@ type_check_rel_op(Env, Op, P, Arg1, Arg2) ->
             end
     end.
 
-type_check_arith_op(Env, Op, P, Arg1, Arg2) ->
+type_check_arith_op(Env, _, _, Arg1, Arg2) ->
     {Ty1, VB1, Cs1} = type_check_expr(Env, Arg1),
     {Ty2, VB2, Cs2} = type_check_expr(Env, Arg2),
 
-    case compat_arith_type(Ty1,Ty2) of
-        false ->
-          throw({type_error, arith_error, Op, P, Ty1, Ty2});
-        Ty ->
-            {Ty
-            ,union_var_binds(VB1, VB2, Env#env.tenv)
-            ,constraints:combine(Cs1, Cs2)}
-    end.
-
-type_check_int_op(Env, Op, P, Arg1, Arg2) ->
+    TNumber = type(number),
+    TFloat = type(float),
+    expect_expr_subtype(Arg1, Ty1, TNumber, #tenv{}),
+    expect_expr_subtype(Arg2, Ty2, TNumber, #tenv{}),
```

Comment:
You're dropping the constraints from `expect_expr_subtype`. They should be collected and combined with the other constraints.